### ISSUE TITLE
Training w/o full game, is it possible?

### ISSUE THREAD

**Issue by merlinpan** [gcp/leela-zero#1732]
Considering the AG paper is mainly themed on "MCTS as a policy improvement operator".  In that sense, is it possible to do training w/o full games?

AKA, just take any board position, and train the policy/value output to match the of the source net + 3200 (or whatever) evals.  This is likely more prone for local minimas, but can some fraction be reasonably used?

**Comment by RavnaBergsndot** [comment]
AZ and LZ are both already doing this.

The "t=1 through all the game" choice means that a training game is not a "full game" as it seems like. The policy gradient is strictly based on the MCTS visits, but the actual move it plays out is fairly random(though it's still guided by the MCTS). Every position in a training game has a rich source of randomness in it, including the board state and history.

**Comment by isty2e** [comment]
It is of course possible in principle, but it becomes much harder to recover from misevaluation. Consider a semeai with many liberties, for example. If the value head considers a losing situation as winning (say, 80%), the MCTS value output will be lower but still off by a lot (like 65%), whereas the actual winrate would be much lower. Thus it becomes much harder to get away from wrong evaluations unless we train the value head with actual results of the games.

**Comment by Ishinoshita** [comment]
@merlinpan : This is a subject that intrigues me as well for a while, so, although I'm no expert, let me please add my two cents, so that I get a chance to get corrected/answered as well.

Regarding the AGZ paper, although it's just plainly written, it took me time to read it correctly and identify the two different improvement processes:
* "MCTS may therefore be viewed as a powerful **policy** improvement operator."
* Self-play with search – using the improved MCTS-based policy to select each move, then using the game winner z as a sample of the value – may be viewed as a powerful **policy evaluation** operator".

Took me a while to realize there is no such thing as a *value improment operator* in itself, but just the *policy improvement operator* from which a *value* is evaluated/learned.

Using a average value Q accumulated at the root after the 3200 visits (or whatever) as a training target for the value head, as a variant of the TD learning type, has been proposed numerous times here, including by myself in #1197 *(at that time, I had failed to understand that training samples were already used multiple times, so the rest of my comment is just crap).* What I actually proposed was a mixing of the game outcome (z) and the average value at root after MCTS search (Q), like (Q+z)/2. @jkiliani explained me that:
> I also thought of value head target mixing, but @gcp said that while it sounds logical, it's been tried and didn't work. I think he must have experimented with network training sufficiently developing Leela that he's right in that.

Then, from other issues, I got that the drawback of training the value towards the action value Q produced by the search is that is would slowdown the learning. E.g. suppose the value head assessed the so called 'bulky five' shape as uncondiotionally alive, thus the corresponding group as alive (whereas it's dead or at least 'killable'). Accounting this group as alive, the Q value may be close to +1, whereas the 'true' value, that will be proved in the course of the game and reflected in its outcome z will be -1. It will take much more time for the value head to learn the 'truth' from Q than from z, because with TD learning scheme, the true value will 'leak' only very slowy backwards from the game end to the position where that shape arise.

I believe this all the more important in the beginning, to bootstrap 'from nothing'. D. Silver has a long record on dynamic programming and TD learning, so I guess he and his co-authors have tried any possible forms of TD approach before the AG paper solution. In that seminal paper, the breakthrough was 'learning a value function, the graal of computer go, and has has been achieved by training on the game outcome z. The following papers AGZ and AZ didn't depart from this, so it's a strong *hint* that other approaches are not as good, a priori.

This been said, I'm still half convinced it's a bad idea for LZ *with the current sttings*, since the project has switched to using randomness (temperature) all along the self-play games:

* In AGZ, t=1 is used only during the 30 first moves, then it's turned off (t=0). Suppose a blunder or a clear sub-optimal move is picked by temperature within these 30 first moves, sealing the game and making it switch from, say, Black has a lead and should win to White now has a lead and should win. Meaning that using 'white =win' as training target for the positions before the blunder would be a wrong target. That may happen, but for the vast mojority of positions in that game, 'white =win' (the game outcome) is a good target.
* LZ currently uses t=1 all along the self-play games. This has been introduced to increase games variety (at the same time a bug on Diriclet noise was fixed). Ever since, most self-play games are sealed by a blunder or at least sub-optimal move pickeb by temperature. In practice, there might be several sub-optimal moves from both sides in the course of the game and, in the extreme, one could say that the outcome z is the result of a random walk. In, practice, this is not that bad but one can say at least that the outcome z is a noisier target with t=1 all along the game than with t=0 to evaluate the policy and train the value head. Once a real blunder is picked, the blundering side often quicly succumb. There is 50% chance that the succumbing side was leading and should have won the game. Thus 50% chance that most of the training samples in that game will recieve a 'wrong' value target.

* Now that LZ value head is very knowlegeable and since we use t=1 all along the game, considering using a mix of Q and z sounds like a reasonable option again.

* Some Oracle's guys mentions benefit of training on mix target lamda.z+(1-lambda).Q over just training on z, in terms of learning speed (with lambda =1 in the beginning of the training, then giving more weigth to Q [Lessons From AlphaZero (part 4): Improving the Training Target](https://medium.com/oracledevs/lessons-from-alphazero-part-4-improving-the-training-target-6efba2e71628). And I understand it's still the vanilla AZ, i.e. t=1 only in the early moves. The assertion that it helps might be questioned because this has only been tried for Connect4, not go. But they mention in the end of  their article that:
>We have begun experimenting with training using both q and z for longer games, and so far the results look promising. It would be fascinating to see this approach applied to larger AlphaZero efforts like Leela Chess Zero.

Indeed, although I cannot retrieve where, I've read on Discord/lc0 that some devs of Leela Chess Zero were actually considering it.

Unfortunately, current LZ training data format does not capture the Q value at the end of the search. Adding it would be great for testing such stuff or for other debug usage, but I guess it impacts many parts of the project, would cause backward incompatibility on training data, etc... Not a piece of cake. Not something you would do unless faced with a major stall, what we currently are not.

**Comment by merlinpan** [comment]
It feels fundamentally there must be a value improvement operator somewhere, because human learning works like that.  When practicing play, we "attempt some playout", then "eval the resulting non terminal position".  The learning there is using that eval to refine the earlier eval.

This is not saying that wouldn't get stuck in local minimas, but some fraction of that may be possible.

Your "slow down the learning" do apply, but the question here is whether the much faster generation of learning examples can make up for it or not.  As for your example, when we get a new version of the value head, we can re-run all previous training examples and update the training set based on the new value head, propagating the value back instead of running full games.

**Comment by Ishinoshita** [comment]
> When practicing play, we "attempt some playout", then "eval the resulting non terminal position"

That's what we do indeed, as human. But the rest of the game and the game outcome may prove our intermediaite evaluation wrong. For instance, you've made a big moyo on one side of the board. That moyo looks really big and your first intuition us that it's a big asset. You consider different variations and conclude that you can convert much of that moyo into solid territory. Thus you evaluate (Q) that you've got a winning position and some winning line of play. The rest of the game proves your moyo to be very thin and successfully invaded by your opponent. You lost that game (z). Then you will have the hard job, as human, to revised the *estimated value* of your initial mid-game position against the actual game outcome. Outcome is ground truth, as long as you cannot find/forced another game trajectory that reverts the outcome. Although that loss, your intuition (value head) is still that that moyo was really big and you *must* have screwed up somewhere. But you don't know where. 
A NN learning on Q value only would be like that human who keeps playing that moyo games because it looks so good, although proved wrong each time. Over the time maybe, that NN will learn to better assess the thickness of a position and true value of the moyo it is making. But slowly, because backwards from the endgame, mid-game, etc... That's the great strength of AGZ/AZ/LZ to be able to learn a value just from the game outcome. Humans usually can't do that and need to rely on a series of intermediate goals that they elaborate for themselves, but which may be flawed. Theses flaws are hard to identify and correct for humans.

NB: My point in favor of considering using a mix of Q and z is *only* in view of the choice of using randomness (temperature) *all along* the self-play games.

> This is not saying that wouldn't get stuck in local minimas

There are an accumulation of theoritical results that tend to say this is probably not an issue. Provided the network is large enough (large redundency in parameters number vs what is to to be learnt ), all minima are very close to global minimum. Saddle points is the issue. However, since we do not use a fix / pre-defined training set, that it is changing over the time (training window), I don't know if this applies to LZ, to be franck.
